### ISSUE TITLE
Switching between channels could be streamlined and result in less downloaded artifacts

### ISSUE THREAD

**Issue by sethladd** [flutter/flutter#152277]
### Use case

We want to encourage folks to try beta channel.

### Proposal

 If it's easy to do, we want to find ways to streamline and simplify.

Today, developers need to do two steps:

```
flutter channel beta
flutter upgrade
```

Both of those options require the user to download artifacts. Seems like there might be duplicative downloading?

One idea here is to add a `flutter channel --upgrade` which does two things:

* runs upgrade when channel switching is done
* smartly downloads artifacts _once_

**Comment by sethladd** [comment]
cc @loic-sharma @jmagman @anderdobo @andrewkolos @kenzieschmoll

**Comment by andrewkolos** [comment]
> flutter channel --upgrade
 
What would this provide over `flutter channel beta && flutter upgrade`?

> Seems like there might be duplicative downloading?

If this is true, this is something we could consider addressing, hopefully without needing a new command option. cc @christopherfujino for possible additional knowledge.

**Comment by sethladd** [comment]
Yeah, maybe we ran into one weird scenario where it looked like artifacts were downloaded twice.

if `flutter channel beta && flutter upgrade` only downloads the sufficient artifacts once, then that would work!

**Comment by loic-sharma** [comment]
> > flutter channel --upgrade
> 
> What would this provide over `flutter channel beta && flutter upgrade`?

The correct equivalent would be `flutter channel beta --no-cache-artifacts && flutter upgrade`. This is very inconvenient to type out :)

**Comment by sethladd** [comment]
Aside, after running `flutter channel beta`, I see "To ensure that you're on the latest build from this channel, run 'flutter upgrade'" 

If the common thing is to always run `flutter upgrade` after a channel switch, maybe that should be the default behavior of a channel switch? (that is, run upgrade for the user, after a channel switch)

**Comment by christopherfujino** [comment]
> Aside, after running `flutter channel beta`, I see "To ensure that you're on the latest build from this channel, run 'flutter upgrade'"
> 
> If the common thing is to always run `flutter upgrade` after a channel switch, maybe that should be the default behavior of a channel switch? (that is, run upgrade for the user, after a channel switch)

Previous conversation: https://github.com/flutter/flutter/issues/28153#issuecomment-577444822

I could have sworn Tim also filed this issue, but I can't find it so yeah.

**Comment by christopherfujino** [comment]
> > Aside, after running `flutter channel beta`, I see "To ensure that you're on the latest build from this channel, run 'flutter upgrade'"
> > If the common thing is to always run `flutter upgrade` after a channel switch, maybe that should be the default behavior of a channel switch? (that is, run upgrade for the user, after a channel switch)
> 
> Previous conversation: [#28153 (comment)](https://github.com/flutter/flutter/issues/28153#issuecomment-577444822)
> 
> I could have sworn Tim also filed this issue, but I can't find it so yeah.

As an addendum, I don't have a very strong opinion here, I think it's a product decision. I would just say that moving backward on a channel is not supported, the user would have to do `git reset --hard 1.0.0` if they wanted to undo their upgrade.

**Comment by christopherfujino** [comment]
Well, we do have `flutter downgrade`, which works once.

**Comment by sethladd** [comment]
here's my instinct, and I'd love to hear what @anderdobo thinks.

* if it's expected that the user should run `upgrade` after every channel switch (as the printed instructions seem to hint?), then a channel switch should run `upgrade` automatically for the user
* there should be a `--no-upgrade` option for channel switch, if for some reason that's important to maintain

This is a low priority issue, but we'd welcome patches for this (I assume). Once we want to make a bigger push for folks to try beta channel, this would bump in priority (as we want to streamline channel switching as much as possible).

Can someone confirm if we expect users to run upgrade after most/every channel switch? It seems like that's the advice we print out after a channel switch: "To ensure that you're on the latest build from this channel, run 'flutter upgrade'"

**Comment by christopherfujino** [comment]
> here's my instinct, and I'd love to hear what @anderdobo thinks.
> 
> * if it's expected that the user should run `upgrade` after every channel switch (as the printed instructions seem to hint?), then a channel switch should run `upgrade` automatically for the user
> * there should be a `--no-upgrade` option for channel switch, if for some reason that's important to maintain
> 
> This is a low priority issue, but we'd welcome patches for this (I assume). Once we want to make a bigger push for folks to try beta channel, this would bump in priority (as we want to streamline channel switching as much as possible).
> 
> Can someone confirm if we expect users to run upgrade after most/every channel switch? It seems like that's the advice we print out after a channel switch: "To ensure that you're on the latest build from this channel, run 'flutter upgrade'"

@craiglabenz do you have opinions about this, especially with regards to bigger customers who have constraints from keeping at the tip of a channel?

**Comment by craiglabenz** [comment]
Large customers with constraints stay on stable - 1, so I don't think they're running this command and will be terrible worried either way.

I like Seth's idea of having `flutter channel x` automatically upgrade with a flag to opt-out. To me, that is the ideal DX.

**Comment by anderdobo** [comment]
Consider the scenario: If I'm intentionally on a specific Flutter SDK version on the stable channel for my day to day development and I switch to beta temporarily to experiment (which we want more of), I don't want my Flutter SDK to be automatically upgraded when I switch back to stable. It would not be obvious that I need to explicitly opt out of upgrades when switching back to stable, unless we prompt for this each time you switch channels.

We don't automatically upgrade the SDK when it is initiated, so IMO to keep things consistent, upgrades should always be explicitly initiated by the developer.

Other than  the unwieldy `flutter channel beta --no-cache-artifacts && flutter upgrade`, (which certainly counts as explicit) a middle ground could be when switching channels, the tool prompts the developer "Do you want to upgrade to the latest available version? [Y/N]" or similar.

**Comment by sethladd** [comment]
>  "Do you want to upgrade to the latest available version? [Y/N]" or similar.

maybe we can help by saying "you are on version/date X, and the latest version/date for this channel is Y. Your [channel] version is  ZZ days old. Do you want to upgrade to the latest available version of [channel]? [y/n]"

that way, folks have some info around "should I update the version of my channel?"

wdyt?

**Comment by craiglabenz** [comment]
Our proposed DX change keeps getting better with every new comment. Part of me wants this thread to run forever so the feature we end up shipping here can solve world hunger, too.

All kidding aside, I love where this thread has landed. My only final note here is to still expose the `--[no-]upgrade` flag so that users can fully automate this workflow in CI.

**Comment by anderdobo** [comment]
I would love to pursue solving world hunger @craiglabenz, but the fun's got to stop some _somewhere_.

Based on above, ideal UX update I think we should assess feasibility of:
1. Adding a prompt that's shown when you switch channel **and** the SDK on that channel you are switching to is not the latest: `You are on version X, and the latest version for this channel is Y. Your [channel] version is ZZ days old. Do you want to upgrade to the latest available version of [channel]? [y/n]`
I shortened slightly by taking out date, as "ZZ days old" conveys the same info (freshness)
2. Adding a  `--[no-]upgrade` flag so that users can fully automate this workflow in CI.

@christopherfujino @andrewkolos how does this look to you?

**Comment by andrewkolos** [comment]
@anderdobo SGTM, and I'm guessing implementing this would be feasible

**Comment by bkonyi** [comment]
Closing as duplicate https://github.com/flutter/flutter/issues/14979
